Dependency Injection should use method injection instead of constructor injection

ObliviousHarmony opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

One of the caveats to using constructor injection is that we must maintain backwards compatibility for constructors in the non-internal sections of our namespace. Given that declaring constructors internal is likely impractical, it seems the best course of action would be to use a consistent method for dependency injection that we can develop a sniff around enforcing the `internal` nature of.

This PR proposes replacing constructor injection with method injection via a `set_internal_dependencies()` method. In this I've added a custom `Definition` override that removes support for constructor injection entirely, replacing it with this method injection. In this way the usage of the container is the same and constructors can remain for initialization of class state that does not require these dependencies.

### How to test the changes in this Pull Request:

1. Run test suite

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Dev - Replaced constructor injection with `set_internal_dependencies()` method injection in order to better support backwards compatibility.

ObliviousHarmony committed:
Fixed comments with incorrect namespaces

ObliviousHarmony committed:
Replaced constructor injection with method injection

Since we need to maintain backwards compatibility for class constructors we should settle on using method injection instead of constructor injection. I've replaced the `Definition` class we're using with one that doesn't support constructor arguments and added a check for auto_arg addition. Note that we don't check for method existence in the extended container. This is because reflection is unnecessarily expensive and we should avoid it if at all possible.

ObliviousHarmony commented on src/Container.php:
```diff
@@ -23,17 +23,12 @@
-	/**
-	 * The root namespace of all WooCommerce classes in the `src` directory.
-	 */
-	const WOOCOMMERCE_ROOT_NAMESPACE = 'Automattic\\WooCommerce';
```
Since this constant is only used in the `ExtendedContainer` it felt like it belonged there. If it was to belong anywhere I would imagine it might be some central `WooCommerce` namespaced class, but for now, a private property seems like a good fit?

Konamiman replied:
Yep, if it's private stuff it's ok, we can move it around later if we need to.

ObliviousHarmony commented on src/Internal/DependencyManagement/Definition.php:
```diff
@@ -0,0 +1,41 @@
+/**
+ * An extension of the definition class that replaces constructor injection with method injection.
+ */
+class Definition extends BaseDefinition {
```
It seemed useful that we entirely disable constructor injection to prevent mistakes. We can use this class in place of thephpleague/container's `Definition` class to accomplish that task.

Konamiman replied:
Sure, but then we need to actually disallow constructors with arguments, see my comment in `ExtendedContainer.php`.

ObliviousHarmony commented on src/Internal/DependencyManagement/ExtendedContainer.php:
```diff
@@ -43,22 +51,21 @@ class ExtendedContainer extends \League\Container\Container {
 
-		return parent::add( $class_name, $concrete, $shared );
-	}
+		$concrete_class = $this->get_class_from_concrete( $concrete );
```
I should note that this isn't entirely comprehensive because they can set a closure to a concrete that returns an invalid class, but given how unlikely that is, I think this should be good enough. I considered adding some error mechanism to `Definition` for the callable case but I felt like that might be overkill?

Konamiman replied:
Using an obscure concrete like that would be akin to using reflection to invoke a private method: not something that any of us would approve in a PR except in very very special cases. So indeed, I think extra error checking isn't necessary.

Konamiman commented on src/Utilities/StringUtility.php:
```diff
@@ -0,0 +1,62 @@
+/**
+ * A class of utilities for dealing with namespaces.
+ */
+final class StringUtility {
```
What would you think about naming this class just `String`? The code would read much better, and the class is behind a namespace so that being a too generic name isn't a problem.

Konamiman replied:
Also just `Utils` instead of `Utilities` as the namespace? (not sure about this one as I'm not a native English speaker, rather I'm an English-abuser)

ObliviousHarmony replied:
`String` is a reserved word but I've made it `Str`, let me know if that works!

As for the `Utilities` I generally prefer the longer form as a namespace, what do you think?

Konamiman commented on tests/php/src/Internal/DependencyManagement/AbstractServiceProviderTest.php:
```diff
@@ -84,47 +85,47 @@ function get_new_dependency_class( DependencyClass $dependency ) {
 
 	/**
-	 * @testdox 'add_with_auto_arguments' should throw an exception if the passed class has a private constructor.
+	 * @testdox 'add_with_auto_arguments' should throw an exception if the passed class has a private method.
```
Better "...has a private injection method". The same for the other tests with a similar comment.

Konamiman commented on tests/php/src/Internal/DependencyManagement/ExampleClasses/ClassWithConstructorArgumentWithoutTypeHint.php:
```diff
@@ -13,10 +13,10 @@
 class ClassWithConstructorArgumentWithoutTypeHint {
```
For consistence we should rename this class to something like `ClassWithSetDependenciesArgumentWithoutTypeHint`

ObliviousHarmony replied:
Thanks, I meant to rename these classes and tests accordingly but it slipped my mind by the time it came to actually do it 🤣

Konamiman commented on tests/php/src/Internal/DependencyManagement/ExtendedContainerTest.php:
```diff
@@ -38,11 +38,23 @@ public function test_add_throws_when_trying_to_register_class_in_forbidden_names
 	}
 
+	/**
+	 * @testdox 'add' should throw an exception when trying to register a concrete class not in the WooCommerce root namespace.
```
Good catch! I hadn't thought of that one.

Konamiman commented on tests/php/src/Proxies/LegacyProxyTest.php:
```diff
@@ -33,7 +33,7 @@ public function setUp() {
 	public function test_get_instance_of_throws_when_trying_to_get_a_namespaced_class() {
 		$this->expectException( \Exception::class );
-		$this->expectExceptionMessage( 'The LegacyProxy class is not intended for getting instances of classes in the src directory, please use constructor injection or the instance of \Psr\Container\ContainerInterface for that.' );
+		$this->expectExceptionMessage( 'The LegacyProxy class is not intended for getting instances of classes in the src directory, please use method injection or the instance of \Psr\Container\ContainerInterface for that.' );
```
Since this error message is intended to be "educational", maybe we could explicitly include the name of the method?

Konamiman commented on src/Internal/DependencyManagement/AbstractServiceProvider.php:
```diff
@@ -78,21 +76,20 @@ protected function add_with_auto_arguments( string $class_name, $concrete = null
 	private function reflect_class_or_callable( string $class_name, $concrete ) {
```
The method comment still has a couple of mentions to the constructor, those should be adjusted.

Konamiman commented on src/Internal/DependencyManagement/ExtendedContainer.php:
```diff
@@ -43,22 +51,21 @@ class ExtendedContainer extends \League\Container\Container {
+			throw new ContainerException( "You cannot add '$class_name', only classes in the {$this->woocommerce_namespace} namespace are allowed." );
 		}
 
-		return parent::add( $class_name, $concrete, $shared );
```
Before doing this we need to check that the class either doesn't have a constructor or has one with no arguments, otherwise " Too few arguments to function __construct" will be thrown when trying to resolve the class.

ObliviousHarmony replied:
I looked at a few options for this:
- `is_callable( array( $class_name, '__construct' ) )` is always `false` due to an idiosyncrasy in PHP.
- `method_exists( $class_name, '__construct' ) )` prevents cases where a constructor may be used to perform initialization such as a `Logger` class that opens a file reference and closes it in `__destruct `.
- We should avoid reflection in the core container wherever possible so we can't check to make sure there's no arguments.

I'm wondering now if my argument against `method_exists()` is invalid given that they can add an inflector with an `init` method. Arguably that may be preferable and it would be easy to add tooling to ensure we never add a constructor to our classes.

Konamiman replied:
I think we need to take a step back here.

We have a `set_whatever_dependencies` method that is executed at class creation time, to set the dependencies. And in some cases there might be a need to perform object initialization when it's created, in a `init`-like routine.

The thing is, aren't these two routines basically the same, and by combining them we get a full constructor replacement? So we can just call the routine `init` (or if you want a less generic name, `init_instance` or the like), have it take any required dependencies, and use it for any required initialization too.

And as a bonus, that would settle the discussion on the method name too.

ObliviousHarmony replied:
Per our conversation we will settle on `init_instance` as a name.

ObliviousHarmony replied:
I'm documenting here that I noticed some classes use an `init_instance_xxx` and felt a name more descriptive to the specific use-case might be preferable. To that end I've proposed `container_init` since it is an initialization initiated by the container.

Konamiman commented on src/Internal/DependencyManagement/ExtendedContainer.php:
```diff
@@ -43,22 +51,21 @@ class ExtendedContainer extends \League\Container\Container {
-	private function class_is_in_root_namespace( $class_name ) {
-		return substr( $class_name, 0, strlen( Container::WOOCOMMERCE_ROOT_NAMESPACE ) + 1 ) === Container::WOOCOMMERCE_ROOT_NAMESPACE . '\\';
+		// We want to use a definition class that does not support constructor injection to avoid accidental usage.
+		if ( ! $concrete instanceof DefinitionInterface ) {
+			$concrete = new Definition( $class_name, $concrete );
+		}
```
There's a "backdoor" here: an instance of `\League\Container\Definition\Definition` could be passed as a concrete, and this would allow to happily register a class with constructor injection. So better something like this:

```
if ( ! $concrete instanceof DefinitionInterface ) {
    $concrete = new Definition( $class_name, $concrete );
}
else if( ! $concrete instanceof Definition ) {
    $concrete = new Definition( $concrete->getAlias(), $concrete->getConcrete() );
}
```

ObliviousHarmony replied:
My concern with checking for the base `Definition` is that cases where it is passed might also include arguments, inflectors, and other things that we aren't actually cloning here. Given that they could also implement `DefinitionInterface` in a custom class if they wanted to get around this limitation, my thinking is that we can consider this case the same as we do passing a closure that returns an invalid class.

Konamiman replied:
Agreed. We are going to reject classes with constructor arguments anyway so not a big issue.

Konamiman commented:
I still think that the name of the method should be just `set_dependencies`. In `set_internal_dependencies` the "internal" part doesn't really provide any useful information (a class has dependencies... period) , and it could even end up being confusing ("Internal dependencies? Are there public dependencies or something?")

ObliviousHarmony commented:
> I still think that the name of the method should be just `set_dependencies`. In `set_internal_dependencies` the "internal" part doesn't really provide any useful information (a class has dependencies... period), and it could even end up being confusing ("Internal dependencies? Are there also public dependencies or something? Then how are those set?")

My thinking here is that `set_dependencies()` might be a method used in an actual class that perhaps inherit from one that uses dependency injection. I felt a more specific name would be a good way avoid accidentally colliding with potential child class methods. Do you think that's valid?

ObliviousHarmony committed:
Replaced the rest of the constructor injection references

ObliviousHarmony committed:
Renamed `set_internal_dependencies` to `container_init`

ObliviousHarmony committed:
Added a requirement that `container_init` methods be marked `final` when registering using reflection

ObliviousHarmony committed:
Added a sniff rule for checking the internal injection method

ObliviousHarmony committed:
Updated the woocommerce/woocommerce-sniffs package to include the new PHPCS rule

ObliviousHarmony committed:
Renamed the `StringUtility` to `StringUtil` to make it easier to work with

Konamiman commented on src/Utilities/StringUtil.php:
```diff
@@ -21,7 +21,7 @@ final class StringUtility {
 	 * @return bool True if the $string starts with $starts_with, false otherwise.
 	 */
-	public static function starts_with( string $string, string $starts_with, bool $case_sensitive = true ): bool {
+	public static function starts( string $string, string $starts_with, bool $case_sensitive = true ): bool {
```
I think `starts_with` and `ends_with` are fine, also that's the name other languages use (https://docs.microsoft.com/en-us/dotnet/api/system.string.startswith, https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/startsWith, https://docs.python.org/3/library/stdtypes.html#str.startswith) so maybe it's better not to deviate from "the norm" here.

rrennick commented:
> I felt a more specific name would be a good way avoid accidentally colliding with potential child class methods. Do you think that's valid?

Child classes can access it through `parent::set_dependencies()` if for some reason they really needed to have their own method `set_dependencies()`.

ObliviousHarmony committed:
Settled on an injection method of `init`

1. Since our code style dictates that these be `final` methods, we shouldn't be concerned about overlap.
2. There is precedent for `init` methods as requirements before using class instances.

ObliviousHarmony committed:
Merge branch 'master' into feature/container-injection-change

ObliviousHarmony committed:
Removed classes accidentally added back by merge from master

ObliviousHarmony committed:
Added `_with` suffix to `StringUtil` methods for clarity

ObliviousHarmony committed:
Removed unnecessary @package tags

ObliviousHarmony commented:
After iterating through a few different names and discussing the matter with @Konamiman, I've settled on `init` for the injection method:
- There is already precedent for classes that have an `init` method to call that method prior to usage.
- Since it must be marked `final` according to our sniff, it shouldn't matter for inheritance and new classes will already have this expectation.

rrennick reviewed (approved):
@ObliviousHarmony Looks good.

jasonsalzwedel commented:
Hi there,

I hope I am not wasting your time here, but I have just picked up a bug in my wordpress website. I am taking a chance that it might be related to the work done here. There error is:

Parse error: syntax error, unexpected ':', expecting '{' in /home/havenrlz/public_html/wp-content/plugins/woocommerce/woocommerce.php on line 56

That section of code is given below, with line 56 being `function wc_get_container() : \Psr\Container\ContainerInterface {`
```
/**
 * Returns the WooCommerce PSR11-compatible object container.
 * Code in the `includes` directory should use the container to get instances of classes in the `src` directory.
 *
 * @return \Psr\Container\ContainerInterface The WooCommerce PSR11 container.
 */
function wc_get_container() : \Psr\Container\ContainerInterface {
	return $GLOBALS['wc_container'];
}
```

Konamiman commented:
Hi @jasonsalzwedel, what versions of PHP, WordPress and WooCommerce are you using?

Konamiman commented:
Hi @jasonsalzwedel, I've been able to reproduce the issue by switching PHP version to 5.6, so I guess that's the version you are running.

Please note that [as of version 3.9 WooCommerce has a minimum requirement of PHP 7.0](https://woocommerce.wordpress.com/2020/01/22/woocommerce-3-9-has-landed/). So if that's the case, you need to upgrade your PHP version to 7.0 or newer (we actually recommend 7.2 or newer). Sorry for the inconveniences.

jasonsalzwedel commented:
Hi @Konamiman, Fantastic. I updated the PHP version and everything is working 100%. I really appreciate the help on this. Thanks!
